README
gadoyanspam - a spam killer for qmail
version 0.2 "Re re re re" - 20040127 GPL
Mohammad DAMT <mdamt@bisnisweb.com>
(c) 2004, PT Cakram Datalingga Duaribu (http://www.cdl2000.com)

This is gadoyanspam, a spam killer for qmail.
USE AT YOUR OWN RISK. NO GUARANTEE AT ALL.

WHAT IS gadoyanspam
gadoyanspam is a spam killer for qmail. It can process a message, 
learn a spam message, and relearn a message incorrectly marked as a spam.
gadoyanspam uses and relies on DSPAM[1] library as the anti spam engine.
This software is sponsored and used by Bisnisweb.NET[2], Linux ASP.Net Web Hosting.

gadoyanspam is a word come from Betawi language, used in Jakarta, Indonesia.
(ka)ga = (I) don't doyan = like gadoyanspam = I don't like spam

The web site is http://oss.mdamt.net/gadoyanspam/
No bugzilla, just report your bugs (or success) to 
gadoyanspam@mdamt.net

This software is known to work in Linux Mandrake 9.2. If you
have gadoyanspam works in another systems, please let me know.

WHAT IS DSPAM
You can find more information about DSPAM in it's web page (url is in the end of
this document). DSPAM's performance and accuracy is amazing. Every day
I used to have more than 50 spams, and now 0.

WHY DO I HAVE TO USE gadoyanspam IF I CAN USE DSPAM?
Good question. I use qmail. DSPAM doesn't support qmail out of the box.
And I don't like the solution to use DSPAM with qmail mentioned in DSPAM's 
website. So I need to use gadoyanspam. If you dont use qmail then dont use gadoyanspam. 
Hey this is _the_ README document! It is not supposed to be an FAQ!

HOW TO USE
* Install DSPAM. I use version 2.8.1
  We dont use the "dspam" program that come with the package, 
  but we need the other executables and libdspam
  from the package. Dont forget to setup the USERDIR.
  READ the documentation

* Compile gadoyanspam:

  # gcc -o gadoyanspam gadoyanspam.c -ldspam -I/path/to/dspam-source-code
  
  If your dspam installation is not in your library path, you can try:

  # gcc -o gadoyanspam gadoyanspam.c -ldspam -I/path/to/dspam-source-code\
    -L/path/to/dspam/lib

* Install it somewhere in your PATH

  # cp gadoyanspam /usr/bin

* Put it in your .qmail file, like:

  |/path/to/gadoyanspam 
  ./Maildir/username

* by default you can send:
  - normal messages to:  user@domain.com
  - spam messages to: user-spam@domain.com
  - false positives to: user-notspam@domain.com

  dont forget to create email user-spam@domain.com and user-notspam@domain.com
  these address dont have to have a mailbox, /dev/null is ok. But if you are
  using quarantine feature, at least user-spam@domain.com needs to have a maildir
  installed.

  read documentation about arguments below to change the address

* maintain your spam database according to dspam documentation

SOME ARGUMENTS
* --deliver-spam:    Force gadoyanspam to pass the spam message through.
                     However, the spam tokens are learned.
* --spam-tag:        Followed by a tag name, indicate what tag appended
                     to the spam box email address. For example, if you have
                     bush-spam@miserable-failure.com as your spam box,
                     then the tag is -spam
					 default: -spam
* --innocent-tag:    Followed by a tag name, indicate what tag appended
                     to the false positive email address. For example,
                     if you have bush-guilty@miserable-failure.com as your
                     false positive email address, then the tag is
                     -guilty
					 default: -notspam
* --force-single:    Force gadoyanspan to use a single database for all
                     email accounts in this domain. Useful if you want to 
					 share a single database across users. Please read 
					 the warning on DSPAM Documentation about groups.
					 In this mode, spam and innocent tags should not
					 prefixed with dash.
					 default: no
* --quarantine:      It specifies where to save suspected spam message.
                     It must be a maildir. gadoyanspam will create the directory
					 if the directory doesn't exists. The directory specified here
					 is relative to $HOME and will be appended with --spam-tag value
					 default: $EXT or --spam-tag value (if --force single is active)
* --no-quarantine:   It specifies whether to save suspected spam message or not.
                     If this option is set, --quarantine will be ignored
                     default: no

WHAT IS $EXT?
Read qmail-command.8 manual page

FALSE POSITIVES AND SPAM BOX
What are false positives and spam box? read the dspam documentation.
In short, a spam box is an address that you forward all of your spams,
and false positive box is an address that you forward all of your
innocent messages which is being incorrectly detected as spams.

WHAT HAPPEN IF A SPAM MESSAGE IS CATCHED?
If you activate quarantine feature (which is activated by default),
the spam messages will be saved in $HOME/quarantine-value-spamtag-value.
For example, if:

--innocent-tag = -impossible
--spam-tag = -guilty
--quarantine = users/$EXT
$HOME = /home/bush
any spam messages catched for email bush@miserable-failure.com
will be saved in /home/bush/users/bush-guilty

If any of the saved messages are not spams, you can forward them to address
bush-impossible@miserable-failure.com.

any spam messages catched for email bush-is@miserable-failure.com
will be saved in /home/bush/users/bush-is-guilty

RETURN VALUES:
* 0 success, 99 spam (0 in --deliver-spam mode), 111 deferral

CAVEAT:
* spam and innocent tags will work only if they are appended to the email address.
  user-spam-john@domain.com and user-notspam-john@domain.com will not work
  if the spam and innocent tags are -spam and -notspam respectively.
  Tags also will not work if spam tag is a substring of innocent tag.
  E.g. this will work: -spam (spam) and -notspam (innocent)
       this won't work: -isjunk (spam) and -junk (innocent)

KNOWN BUGS:
* gadoyanspam doesn't know about groups in dspam USERDIR. However, gadoyanspam
  have similar feature using --force-single flag.

TO ALL NIGERIAN AND LOTTO SCAMMERS, SEX SHOPS, PORN SITES, DRUG STORES, etc:
Please send your proposals and offerings to spam@mdamt.net.
And please consider to use gadoyanspam in your mail servers,
|t w|(( S+T+O+P [S/P/A/M] 4s s.e.e.n 0n CnN,f0x,C3S, e(v)en 0pR4h!!

+---notes for vmailmgr users:-------------------------------------------
|* Install vmailmgr[3]. I use version 0.96.9
|  patch it using vmailmgr-patch provided by gadoyanspam
|
|* add these lines below to /etc/vmailmgr/vdeliver-predeliver
|  if you dont have the file, create it. Note: gadoyanspam
|  will save suspected message in $HOME/users/$EXT directory.
|  Modify the --quarantine to fits your vmailmgr setting.
|
| [---START CUT HERE---]
| #!/bin/bash
| /path/to/gadoyanspam --innocent-tag=-notspam --spam-tag=-spam\
| --quarantine=users/$EXT
|
| ## uncomment these lines below if you have vcheckquota
| #ret=$?
| #if [ $ret -ne 0 ]; then
| #	exit $ret
| #fi
| #
| #/usr/local/bin/vcheckquota
| [---END TO HERE---]
|
|* dont forget to mark /etc/vmailmgr/vdeliver-predeliver as executable
|
|  # chmod +x /etc/vmailmgr/vdeliver-predeliver
+-----------------------------------------------------------------------

[1] http://www.nuclearelephant.com/projects/dspam/
[2] http://www.bisnisweb.net/
[3] http://www.vmailmgr.org/


ChangeLog
2004-01-27	Mohammad DAMT  <mdamt@bisnisweb.com>
	
	* ChangeLog started
	* removed --vmailmgr option. We need to patch vmailmgr to work with gadoyanspam
	* added --force-single to use single database for all virtual accounts
	* added --quarantine and --no-quarantine options
	* updated README
	* version 0.2


gadoyanspam.c
/*

gadoyanspam - a spam killer for qmail.
Mohammad DAMT [mdamt at bisnisweb dot com] 
(c) 2004, PT Cakram Datalingga Duaribu
   
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <ctype.h>

#include "libdspam.h"

#define SPAM_TAG "-spam"
#define INNOCENT_TAG "-notspam"

int quarantine = 1;
int force_single = 0;
int drop_spam = 1;

void die_fail ()
{
	if (drop_spam) {
		exit (99);
	} else 
		exit (0);
}

void die_temp (char *message)
{
	fprintf(stderr, message);
	exit(111);
}

void save_maildir (char *path, char *message)
{
	time_t now = time (NULL);
	char hostname [512];
	pid_t pid = getpid ();
	char *filename, *new;
	int filename_len;
	struct stat st;
	int fd;
	int ret;

	bzero (hostname, 512);
	if (gethostname (hostname, 512) != 0)
		strcpy (hostname, "localhost");

	mkdir (path, 0700);
	if (chdir (path) != 0)
		die_temp ("couldnt chdir() to maildir path");

	mkdir ("tmp", 0700);
	mkdir ("new", 0700);
	mkdir ("cur", 0700);

	for (;;sleep (2)) {
		filename_len =	4 + /* tmp/ */
						strlen (hostname) +
						20 + /* now */
						10 + /* pid */
						4; /* \0 */

		filename = malloc (filename_len);
		bzero (filename, filename_len);
		snprintf (filename, filename_len, "tmp/%d.%d.%s", (unsigned int) now, (int) pid, hostname);

		if (stat (filename, &st) == 0)
			continue;

		if ((fd = creat (filename, 0600)) == -1)
			die_temp ("couldn't create quarantine message");
		
		
		if ((ret = write (fd, message, strlen (message))) < 0) {
			die_temp ("couldn't write quarantine message");
		}

		fsync (fd);
		close (fd);
		break;
	}
	
	new = strdup (filename);
	if (new == NULL)
		die_temp ("no memory for new/");

	new [0] = 'n'; new [1] = 'e'; new [2] = 'w';

	link (filename, new);

	unlink (filename);
	free (filename);
	free (new);
}

int main (int argc, char **argv) 
{
	char buffer [1024];
	char *message = malloc (1);
	char *home = getenv ("HOME");
	char *spam_tag = NULL, *innocent_tag = NULL;
	char *local = NULL, *temp, *quarantine_path = NULL;
	char *sublocal = NULL;
	int local_len = 0;
	long len = 1;
	int result = -1, i = 0;
	int addspam = 0;

	DSPAM_CTX *CTX;

	while (1) {
		int option_index = 0;
		static struct option long_options[] = {
			{"spam-tag", 1, 0, 0},
			{"innocent-tag", 1, 0, 0},
			{"quarantine", 1, 0, 0},
			{"no-quarantine", 0, 0, 0},
			{"deliver-spam", 0, 0, 0},
			{"force-single", 0, 0, 0},
			{0, 0, 0, 0}
		};

		i = getopt_long (argc, argv, "", long_options, &option_index);
		if (i == -1)
		break;

		if (i == 0) {
			if (strcmp ("spam-tag", long_options[option_index].name) == 0) {
				if (optarg)
					spam_tag = strdup (optarg);
			}
			else if (strcmp ("innocent-tag", long_options[option_index].name) == 0) {
				if (optarg)
					innocent_tag = strdup (optarg);
			}
			else if (strcmp ("quarantine", long_options[option_index].name) == 0) {
				if (optarg)
					quarantine_path = strdup (optarg);
			}
			else if (strcmp ("deliver-spam", long_options[option_index].name) == 0) {
				drop_spam = 0;
			}
			else if (strcmp ("force-single", long_options[option_index].name) == 0) {
				force_single = 1;
			}
			else if (strcmp ("no-quarantine", long_options[option_index].name) == 0) {
				quarantine = 1;
			}
		}
	}
	
	if (spam_tag == NULL)
		spam_tag = strdup (SPAM_TAG);

	if (spam_tag == NULL) {
		if (innocent_tag)
			free (innocent_tag);
		die_temp ("spam-tag is empty");
	}
	
	if (innocent_tag == NULL)
		innocent_tag = strdup (INNOCENT_TAG);
	
	if (innocent_tag == NULL) {
		if (spam_tag)
			free (spam_tag);
		die_temp ("innocent-tag is empty");
	}
			
	if (strcmp (spam_tag, innocent_tag) == 0) {
		die_temp ("SPAM and FALSE POSITIVE can't have the same tag");
	}
	
	if (home == NULL)
		die_temp ("$HOME is not set\n");

	if (getenv ("EXT") == NULL)
		die_temp ("$EXT is not set\n");

	if ((temp = getenv ("LOCAL")) == NULL)
		die_temp ("$LOCAL is not set\n");
	
	local = strdup (temp);
	
	if (local == NULL)
		die_temp ("out of memory #0\n");

	temp = local;
	len = strlen (temp);
	if (len <= 0)
		die_temp ("$LOCAL is not set\n");
	
	while (i ++ < len) {
		temp [i] = tolower (temp [i]);
	}
	
	if ((sublocal = strstr (local, innocent_tag)) != NULL) {
		if (force_single == 0) {
			local_len = strlen (local) - strlen (sublocal) + 1;
			local = malloc (local_len + 1);
			if (local == NULL)
				die_temp ("out of memory #sublocal innocent_tag\n");

			bzero (local, local_len + 1);
			snprintf (local, local_len, "%s", temp);
		}
		addspam = -1;
		if (temp)
			free (temp);
	} else if ((sublocal = strstr (local, spam_tag)) != NULL) {
		if (force_single == 0) {
			local_len = strlen (local) - strlen (sublocal) + 1;
			local = malloc (local_len + 1);
			if (local == NULL)
				die_temp ("out of memory #sublocal spam_tag\n");

			bzero (local, local_len + 1);
			snprintf (local, local_len, "%s", temp);
		}
		addspam = 1;
		if (temp)
			free (temp);
	}	

	if (force_single == 1) {
		local = strdup (getenv ("USER"));
		if (local == NULL) 
			die_temp ("out of memory #force_single");
	} 

	if (quarantine) {
		if (force_single) {
			if (quarantine_path == NULL)
				quarantine_path = strdup (spam_tag);

			if (quarantine_path == NULL)
				die_temp ("out of memory\n");

			temp = quarantine_path;

			i = strlen (home) + strlen (temp) + 2;
			quarantine_path = malloc (i);

			if (quarantine_path == NULL)
				die_temp ("out of memory\n");

			sprintf (quarantine_path, "%s/%s", home, temp);
			quarantine_path [i] = 0;
		} else {
			if (quarantine_path == NULL)
				quarantine_path = strdup (getenv ("EXT"));

			if (quarantine_path == NULL)
				die_temp ("out of memory\n");

			temp = quarantine_path;

			i = strlen (home) + strlen (temp) + strlen (spam_tag) + 2;
			quarantine_path = malloc (i);

			if (quarantine_path == NULL)
				die_temp ("out of memory\n");

			sprintf (quarantine_path, "%s/%s%s", home, temp, spam_tag);
			quarantine_path [i] = 0;
		}
	}

	free (innocent_tag);
	free (spam_tag);

	message [0] = 0;
	while (fgets (buffer, sizeof (buffer), stdin) != NULL) {
		len += strlen (buffer);
		message = realloc (message, len);
		if (message == NULL) 
			die_temp ("out of memory #message\n");
		strcat(message, buffer);
	}

	dspam_init_driver ();
	if (addspam == 1) 
		CTX = dspam_init(local, NULL, DSM_ADDSPAM, DSF_CHAINED | DSF_IGNOREHEADER);
	else if (addspam == -1)
		CTX = dspam_init(local, NULL, DSM_FALSEPOSITIVE, DSF_CHAINED | DSF_IGNOREHEADER);
	else
		CTX = dspam_init(local, NULL, DSM_PROCESS, DSF_CHAINED);

	if (local)
		free (local);
	
	if (CTX == NULL) {
		die_temp ("failure initializing dspam\n");
	}

	if (dspam_process(CTX, message)!=0) 
		die_temp ("dspam_process failed\n"); 

	result = CTX->result;

	if (dspam_destroy(CTX)!=0) 
		die_temp ("dspam_destroy failed!\n");
	
	if (addspam) {
		free (message);
		return 0;
	}

	if (result == DSR_ISSPAM) {
		if (quarantine) {
			save_maildir (quarantine_path, message);
			free (quarantine_path);
		}
		die_fail ();
	}

	free (message);

	return 0;
}
